Show alert when user leaves

iAmAshuSahoo opened the pull request:
Date: `28 Apr 2024`

Developer Name: `Ashutosh Sahoo`

---

## Issue Ticket Number:-

[Alert should be shown](https://status.realdevsquad.com/tasks/6vbZlXNAvm69fKMrvjOR)

## Description:

1. Alert should be shown when the stream is running and the user tries to navigate to a different page.
2. Alert needs to be shown when the user is trying to close the window/tab where live is running.

Is Under Feature Flag

- [x] Yes
- [ ] No

Database changes

- [ ] Yes
- [x] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)

- [ ] Yes
- [x] No

Is Development Tested?

- [x] Yes
- [ ] No

Tested in staging?

- [ ] Yes
- [x] No

### Add relevant Screenshot below ( e.g test coverage etc. )

Test on firefox
![image](https://github.com/Real-Dev-Squad/website-www/assets/7929893/b5ff575a-c27c-4753-a717-09db3df09890)

Test on Chrome
![image](https://github.com/Real-Dev-Squad/website-www/assets/7929893/0e456d90-d35d-4c22-a561-5935e63ac6ba)

Test Coverage
![image](https://github.com/Real-Dev-Squad/website-www/assets/7929893/6f2ba330-9b16-46a2-9588-72c9b2e17c9b)

Build
![image](https://github.com/Real-Dev-Squad/website-www/assets/7929893/893a6f93-7232-4306-b6f8-efae2ab7eb9b)

iAmAshuSahoo commented:
Ways taken for solving that DIDN'T worked

1. Adding window.addeventlistener in constructor of Controller, Component.
```
constructor() {
    super(...arguments);
    window.addEventListener('beforeunload', this.confirmLeave);
  }
willDestroy() {
    window.removeEventListener('beforeunload', this.confirmLeave);
    super.willDestroy(...arguments);
  }
@action
  confirmLeave(event) {
    event.preventDefault();
    event.returnValue = this.confirmationMessage;
    return this.confirmationMessage;
  }
```
Got error - `TypeError: window.addEventListener is not a function`

2. Added the same logic to Route as well with init and  didInsertElement.
3. Using confirmation as a service and invoking this in hbs.
4. `<div  id="live" data-test-live onbeforeunload="return 'Are you sure you want to leave? Your changes may not be saved.';">`
5. Tried adding the addon `ember-onbeforeunload` that uses the concepts of mixin. This is depcrecated

palakgupta2712 commented:
@iAmAshuSahoo 
1. Kindly update the PR description.
2. Please check CI is failing.

mehra-sourav reviewed (commented):
Please add tests

tejaskh3 commented:
kindly add the date and developer name to the PR.
![image](https://github.com/Real-Dev-Squad/website-www/assets/98630752/47b8b021-c9f8-4739-8b6b-f417b60aacf2)

tejaskh3 reviewed (commented):
![image](https://github.com/Real-Dev-Squad/website-www/assets/98630752/62acca77-f817-4ecc-b36e-3fc82eea06ee)
Have you worked on this task before, then please add reference to those previous PRs and add staging proof too. if not then please mark this as "NO"

iAmAshuSahoo commented:
> @iAmAshuSahoo
> 
>     1. Kindly update the PR description.
> 
>     2. Please check CI is failing.

1. Updated description
2. CI is checked.
3. Added tests.
4. updated PR information

satyam73 commented on tests/acceptance/live-test.js:
```diff
@@ -24,24 +24,35 @@ module('Acceptance | live', function (hooks) {
-    */
+    await click('button');
+    assert.strictEqual(currentURL(), '/live', 'we are not on live page');
   });
```
I don't think this test is testing the flow of this PR

iAmAshuSahoo replied:
You are correct. I discussed the implementation of the test for alert in qunit with Ankush. This is not straight-forward and has a learning curve. 

I am asked to close this story and create a new story where developer needs to write a test for alert.

satyam73 commented on app/services/live.js:
```diff
@@ -70,11 +70,26 @@ export default class LiveService extends Service {
   }
 
+  onCloseAlert(event) {
+    console.log('Before Unload Starts');
```
remove console logs

iAmAshuSahoo replied:
removed the logs.

satyam73 replied:
You've removed comments instead of logs

iAmAshuSahoo replied:
I have removed all dead code now (comments and logs)

satyam73 commented on tests/acceptance/live-test.js:
remove this test since you're taking test as separate ticket, please also attach test ticket here

iAmAshuSahoo replied:
removed the code that I added. Some comments were already there I haven't removed. I don't know how to create a ticket. Will raise one with someone's help on working days and add the ticket here.

satyam73 reviewed (approved):
LGTM
